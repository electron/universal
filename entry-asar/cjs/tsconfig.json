{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": ".",
  },
  "include": [
    ".",
    "../ambient.d.ts"
  ],
  "exclude": []
}
