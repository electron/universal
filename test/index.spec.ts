import { spawn } from '@malept/cross-spawn-promise';
import * as fs from 'fs-extra';
import * as path from 'path';

import { makeUniversalApp } from '../dist/cjs/index';

const appsPath = path.resolve(__dirname, 'fixtures', 'apps');
const appsOutPath = path.resolve(__dirname, 'fixtures', 'apps', 'out');

async function ensureUniversal(app: string) {
  const exe = path.resolve(app, 'Contents', 'MacOS', 'Electron');
  const result = await spawn(exe);
  expect(result).toContain('arm64');
  const result2 = await spawn('arch', ['-x86_64', exe]);
  expect(result2).toContain('x64');
}

// See `jest.setup.ts` for app fixture setup process
describe('makeUniversalApp', () => {
  afterEach(async () => {
    await fs.emptyDir(appsOutPath);
  });

  it('throws an error if asar is only detected in one arch', async () => {
    const out = path.resolve(appsOutPath, 'Error.app');
    await expect(
      makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64NoAsar.app'),
        outAppPath: out,
      }),
    ).rejects.toThrow(
      'Both the x64 and arm64 versions of your application need to have been built with the same asar settings (enabled vs disabled)',
    );
  });

  it.todo('works for lipo binary resources');

  describe('force', () => {
    it('throws an error if `out` bundle already exists and `force` is `false`', async () => {
      const out = path.resolve(appsOutPath, 'Error.app');
      await fs.mkdirp(out);
      await expect(
        makeUniversalApp({
          x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
          arm64AppPath: path.resolve(appsPath, 'Arm64Asar.app'),
          outAppPath: out,
        }),
      ).rejects.toThrow(/The out path ".*" already exists and force is not set to true/);
    });

    it('packages successfully if `out` bundle already exists and `force` is `true`', async () => {
      const out = path.resolve(appsOutPath, 'Error.app');
      await fs.mkdirp(out);
      await makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64Asar.app'),
        outAppPath: out,
        force: true,
      });
      await ensureUniversal(out);
      // Only a single asar as they were identical
      expect(
        (await fs.readdir(path.resolve(out, 'Contents', 'Resources'))).filter((p) =>
          p.endsWith('asar'),
        ),
      ).toEqual(['app.asar']);
    }, 60000);
  });

  describe('asar mode', () => {
    it('should correctly merge two identical asars', async () => {
      const out = path.resolve(appsOutPath, 'MergedAsar.app');
      await makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64Asar.app'),
        outAppPath: out,
      });
      await ensureUniversal(out);
      // Only a single asar as they were identical
      expect(
        (await fs.readdir(path.resolve(out, 'Contents', 'Resources'))).filter((p) =>
          p.endsWith('asar'),
        ),
      ).toEqual(['app.asar']);
    }, 60000);

    it('should create a shim if asars are different between architectures', async () => {
      const out = path.resolve(appsOutPath, 'ShimmedAsar.app');
      await makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64AsarExtraFile.app'),
        outAppPath: out,
      });
      await ensureUniversal(out);
      // We have three asars including the arch-agnostic shim
      expect(
        (await fs.readdir(path.resolve(out, 'Contents', 'Resources')))
          .filter((p) => p.endsWith('asar'))
          .sort(),
      ).toEqual(['app.asar', 'app-x64.asar', 'app-arm64.asar'].sort());
    }, 60000);

    it('should merge two different asars when `mergeASARs` is enabled', async () => {
      const out = path.resolve(appsOutPath, 'MergedAsar.app');
      await makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64AsarExtraFile.app'),
        outAppPath: out,
        mergeASARs: true,
        singleArchFiles: 'extra-file.txt',
      });
      await ensureUniversal(out);
      // Only a single merged asar
      expect(
        (await fs.readdir(path.resolve(out, 'Contents', 'Resources'))).filter((p) =>
          p.endsWith('asar'),
        ),
      ).toEqual(['app.asar']);
    }, 60000);

    it('throws an error if `mergeASARs` is enabled and `singleArchFiles` is missing a unique file', async () => {
      const out = path.resolve(appsOutPath, 'Error.app');
      await expect(
        makeUniversalApp({
          x64AppPath: path.resolve(appsPath, 'X64Asar.app'),
          arm64AppPath: path.resolve(appsPath, 'Arm64AsarExtraFile.app'),
          outAppPath: out,
          mergeASARs: true,
          singleArchFiles: 'bad-rule',
        }),
      ).rejects.toThrow(/Detected unique file "extra-file\.txt"/);
    }, 60000);

    it.todo('should not inject ElectronAsarIntegrity into `infoPlistsToIgnore`');
  });

  describe('no asar mode', () => {
    it('should correctly merge two identical app folders', async () => {
      const out = path.resolve(appsOutPath, 'MergedNoAsar.app');
      await makeUniversalApp({
        x64AppPath: path.resolve(appsPath, 'X64NoAsar.app'),
        arm64AppPath: path.resolve(appsPath, 'Arm64NoAsar.app'),
        outAppPath: out,
      });
      await ensureUniversal(out);
      // Only a single app folder as they were identical
      expect(
        (await fs.readdir(path.resolve(out, 'Contents', 'Resources'))).filter((p) =>
          p.startsWith('app'),
        ),
      ).toEqual(['app']);
    }, 60000);

    it.todo('should shim two different app folders');
  });

  // TODO: Add tests for
  // * different app dirs with different macho files
  // * identical app dirs with universal macho files
});
