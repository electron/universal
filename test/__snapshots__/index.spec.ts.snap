// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeUniversalApp asar mode should correctly merge two identical asars 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
        ],
        "hash": "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
      },
      "size": 64,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should correctly merge two identical asars 2`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "85fff474383bd8df11cd9c5784e8fcd1525af71ff140a8a882e1dc9d5b39fcbf",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should create a shim if asars are different between architectures 1`] = `
{
  "files": {
    "extra-file.txt": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "b8f261b95f81761658c8875b33a68001d8750fd898f447373bf6347e779bc3de",
        ],
        "hash": "b8f261b95f81761658c8875b33a68001d8750fd898f447373bf6347e779bc3de",
      },
      "size": 15,
    },
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
        ],
        "hash": "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
      },
      "size": 64,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should create a shim if asars are different between architectures 2`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
        ],
        "hash": "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
      },
      "size": 64,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should create a shim if asars are different between architectures 3`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "b7e5f58d3c0fddc1a57d1279a7f19a34a01784f4036920d4b60a1e33f6d1635b",
        ],
        "hash": "b7e5f58d3c0fddc1a57d1279a7f19a34a01784f4036920d4b60a1e33f6d1635b",
      },
      "size": 1068,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
        ],
        "hash": "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
      },
      "size": 33,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should create a shim if asars are different between architectures 4`] = `
{
  "Contents/Info.plist": {
    "Resources/app-arm64.asar": {
      "algorithm": "SHA256",
      "hash": "71db54541357128943df64d54480a22d0cdf4c283f2044f48101fb1fc6e6fb2d",
    },
    "Resources/app-x64.asar": {
      "algorithm": "SHA256",
      "hash": "85fff474383bd8df11cd9c5784e8fcd1525af71ff140a8a882e1dc9d5b39fcbf",
    },
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "b62aaaed07ff72dc33da1720d900e0443c060285ef374ce1bdaef1d4f28b5fe4",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should generate AsarIntegrity for all asars in the application 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should generate AsarIntegrity for all asars in the application 2`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should generate AsarIntegrity for all asars in the application 3`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "7e6af4d00f4cc737eff922e2b386128a269f80887b79a011022f1276bdbe7832",
    },
    "Resources/webbapp.asar": {
      "algorithm": "SHA256",
      "hash": "7e6af4d00f4cc737eff922e2b386128a269f80887b79a011022f1276bdbe7832",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should merge two different asars when \`mergeASARs\` is enabled 1`] = `
{
  "files": {
    "extra-file.txt": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "b8f261b95f81761658c8875b33a68001d8750fd898f447373bf6347e779bc3de",
        ],
        "hash": "b8f261b95f81761658c8875b33a68001d8750fd898f447373bf6347e779bc3de",
      },
      "size": 15,
    },
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
        ],
        "hash": "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
      },
      "size": 64,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should merge two different asars when \`mergeASARs\` is enabled 2`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "71db54541357128943df64d54480a22d0cdf4c283f2044f48101fb1fc6e6fb2d",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should not inject ElectronAsarIntegrity into \`infoPlistsToIgnore\` 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should not inject ElectronAsarIntegrity into \`infoPlistsToIgnore\` 2`] = `
{
  "Contents/Info.plist": undefined,
  "Contents/Resources/SubApp-1.app/Contents/Info.plist": undefined,
}
`;

exports[`makeUniversalApp asar mode should shim asars with different unpacked dirs 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                  "unpacked": true,
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
              "unpacked": true,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
      "unpacked": true,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should shim asars with different unpacked dirs 2`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
    },
  },
}
`;

exports[`makeUniversalApp asar mode should shim asars with different unpacked dirs 3`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "b7e5f58d3c0fddc1a57d1279a7f19a34a01784f4036920d4b60a1e33f6d1635b",
        ],
        "hash": "b7e5f58d3c0fddc1a57d1279a7f19a34a01784f4036920d4b60a1e33f6d1635b",
      },
      "size": 1068,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
        ],
        "hash": "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
      },
      "size": 33,
    },
  },
}
`;

exports[`makeUniversalApp asar mode should shim asars with different unpacked dirs 4`] = `
[
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
]
`;

exports[`makeUniversalApp asar mode should shim asars with different unpacked dirs 5`] = `
{
  "Contents/Info.plist": {
    "Resources/app-arm64.asar": {
      "algorithm": "SHA256",
      "hash": "d06a628e759f54def7ff8785a077b3a3d756882cb84ee99e9725966226e1f195",
    },
    "Resources/app-x64.asar": {
      "algorithm": "SHA256",
      "hash": "7e6af4d00f4cc737eff922e2b386128a269f80887b79a011022f1276bdbe7832",
    },
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "b62aaaed07ff72dc33da1720d900e0443c060285ef374ce1bdaef1d4f28b5fe4",
    },
  },
}
`;

exports[`makeUniversalApp force packages successfully if \`out\` bundle already exists and \`force\` is \`true\` 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
        ],
        "hash": "8c8cefe616b330a70980c457e479360417a320f53f484d34df65227ce3add026",
      },
      "size": 64,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
  },
}
`;

exports[`makeUniversalApp force packages successfully if \`out\` bundle already exists and \`force\` is \`true\` 2`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "85fff474383bd8df11cd9c5784e8fcd1525af71ff140a8a882e1dc9d5b39fcbf",
    },
  },
}
`;

exports[`makeUniversalApp no asar mode different app dirs with different macho files (shim and lipo) 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
        ],
        "hash": "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
      },
      "size": 1063,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
        ],
        "hash": "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
      },
      "size": 33,
    },
  },
}
`;

exports[`makeUniversalApp no asar mode different app dirs with different macho files (shim and lipo) 2`] = `
[
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with different macho files (shim and lipo) 3`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with different macho files (shim and lipo) 4`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/hello-world.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with different macho files (shim and lipo) 5`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "<stripped>",
    },
  },
}
`;

exports[`makeUniversalApp no asar mode different app dirs with universal macho files (shim but don't lipo) 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
        ],
        "hash": "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
      },
      "size": 1063,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
        ],
        "hash": "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
      },
      "size": 33,
    },
  },
}
`;

exports[`makeUniversalApp no asar mode different app dirs with universal macho files (shim but don't lipo) 2`] = `
[
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with universal macho files (shim but don't lipo) 3`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with universal macho files (shim but don't lipo) 4`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/hello-world.bin",
]
`;

exports[`makeUniversalApp no asar mode different app dirs with universal macho files (shim but don't lipo) 5`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "<stripped>",
    },
  },
}
`;

exports[`makeUniversalApp no asar mode identical app dirs with different macho files (e.g. do not shim, but still lipo) 1`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
]
`;

exports[`makeUniversalApp no asar mode identical app dirs with different macho files (e.g. do not shim, but still lipo) 2`] = `
{
  "Contents/Info.plist": {},
}
`;

exports[`makeUniversalApp no asar mode identical app dirs with universal macho files (e.g., do not shim, just copy x64 dir) 1`] = `
[
  "hello-world",
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
]
`;

exports[`makeUniversalApp no asar mode identical app dirs with universal macho files (e.g., do not shim, just copy x64 dir) 2`] = `
{
  "Contents/Info.plist": {},
}
`;

exports[`makeUniversalApp no asar mode should correctly merge two identical app folders 1`] = `
[
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
]
`;

exports[`makeUniversalApp no asar mode should correctly merge two identical app folders 2`] = `
{
  "Contents/Info.plist": {},
}
`;

exports[`makeUniversalApp no asar mode should shim two different app folders 1`] = `
{
  "files": {
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
        ],
        "hash": "f1e14240f7c833900fca84fabc2f0ff27084efdf1c5b228b015515de3f8fa28e",
      },
      "size": 1063,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
        ],
        "hash": "2873266521e41d58d02e7acfbbbdb046edfa04b6ce262b8987de8e8548671fc7",
      },
      "size": 33,
    },
  },
}
`;

exports[`makeUniversalApp no asar mode should shim two different app folders 2`] = `
[
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode should shim two different app folders 3`] = `
[
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/i-aint-got-no-rhythm.bin",
]
`;

exports[`makeUniversalApp no asar mode should shim two different app folders 4`] = `
[
  "index.js",
  {
    "content": "{
  "name": "app",
  "main": "index.js"
}",
    "name": "package.json",
  },
  {
    "content": "hello world",
    "name": "private/var/file.txt",
  },
  "private/var/hello-world.bin",
]
`;

exports[`makeUniversalApp no asar mode should shim two different app folders 5`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "27433ee3e34b3b0dabb29d18d40646126e80c56dbce8c4bb2adef7278b5a46c0",
    },
  },
}
`;

exports[`makeUniversalApp works for lipo binary resources 1`] = `
{
  "files": {
    "hello-world": "<stripped>",
    "index.js": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
        ],
        "hash": "0f6311dac07f0876c436ce2be042eb88c96e17eaf140b39627cf720dd87ad5b8",
      },
      "size": 66,
    },
    "package.json": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
        ],
        "hash": "d6226276d47adc7aa20e6c46e842e258f5157313074a035051a89612acdd6be3",
      },
      "size": 41,
    },
    "private": {
      "files": {
        "var": {
          "files": {
            "app": {
              "files": {
                "file.txt": {
                  "link": "private/var/file.txt",
                },
              },
            },
            "file.txt": {
              "integrity": {
                "algorithm": "SHA256",
                "blockSize": 4194304,
                "blocks": [
                  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
                ],
                "hash": "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
              },
              "size": 11,
            },
          },
        },
      },
    },
    "var": {
      "link": "private/var",
    },
  },
}
`;

exports[`makeUniversalApp works for lipo binary resources 2`] = `[]`;

exports[`makeUniversalApp works for lipo binary resources 3`] = `
[
  "hello-world",
]
`;

exports[`makeUniversalApp works for lipo binary resources 4`] = `
{
  "Contents/Info.plist": {
    "Resources/app.asar": {
      "algorithm": "SHA256",
      "hash": "<stripped>",
    },
  },
}
`;
